### DISCUSSION TITLE
Working docker image of Theia with python (or how to create it)

### DISCUSSION THREAD

**Discussion by leandrocm86** [eclipse-theia/theia#14160] [Q&A]
Hi. I've tried the latest docker image of Theia I could find: ghcr.io/eclipse-theia/theia-blueprint/theia-ide:1.52.0
However, it seems it doesn't come with python installed. I tried creating a Dockerfile to customize that image and have the python3 interpreter and extension already installed, but couldn't make it work.
Can anyone give me a hand with this, please? Since python is so popular, I feel I'm probably missing something that should have been obvious.

**Comment by msujew** [comment]
Hey @leandrocm86,

> but couldn't make it work.

Can you specify a bit what didn't work and what you already tried?

**Reply by leandrocm86** [reply]
> Hey @leandrocm86,
> 
> > but couldn't make it work.
> 
> Can you specify a bit what didn't work and what you already tried?

Here is the Dockerfile I've tried making myself:

```
FROM ghcr.io/eclipse-theia/theia-blueprint/theia-ide:1.52.0                                                           
                                                                                                                      
# Switch to root user                                                                                                 
USER root                                                                                                             
                                                                                                                      
# Install Python                                                                                                
RUN apt-get update && apt-get install -y python3 python3-pip                                                                                                                  
                                                                                                                      
# Install Python extension for Theia                                                                                 
RUN npm install @theia/python                                                                                         
                                                                                                                      
# Clean up                                                                                                            
RUN apt-get clean && rm -rf /var/lib/apt/lists/*
```

It's failing at the npm step (installing the extension).
Since the error output is too large to paste inline, I'm attatching the log file of the execution:
[theia.log](https://github.com/user-attachments/files/16997955/theia.log)

**Reply by msujew** [reply]
Alright, thanks a lot, the context is really helpful!

> It's failing at the npm step (installing the extension).

I would hope it fails the installation - that's not how you install the extension in the first place. Nevermind that the [`@theia/python`](https://www.npmjs.com/package/@theia/python) hasn't been updated in **5 years**, and is not in use anymore.

If you need Python support in your Theia IDE, you should install the [python](https://open-vsx.org/extension/ms-python/python) and [pyright](https://open-vsx.org/extension/ms-pyright/pyright) extensions in there. You can install them as builtins by placing the extensions in the [`/home/theia/plugins`](https://github.com/eclipse-theia/theia-blueprint/blob/7bac5334764bc9e840bc4c896dba66ae703ad1b9/browser.Dockerfile#L59) directory.

**Reply by leandrocm86** [reply]
> I would hope it fails the installation - that's not how you install the extension in the first place. Nevermind that the [`@theia/python`](https://www.npmjs.com/package/@theia/python) hasn't been updated in **5 years**, and is not in use anymore.
> 
> If you need Python support in your Theia IDE, you should install the [python](https://open-vsx.org/extension/ms-python/python) and [pyright](https://open-vsx.org/extension/ms-pyright/pyright) extensions in there. You can install them as builtins by placing the extensions in the [`/home/theia/plugins`](https://github.com/eclipse-theia/theia-blueprint/blob/7bac5334764bc9e840bc4c896dba66ae703ad1b9/browser.Dockerfile#L59) directory.

Oh, that explains a lot! Thank you for the explanation!
So I suppose I could either mount the plugins folder on a persistent volume or bind-mount it on the host, then just proceed with the regular plugin installation through the UI, right? That way, the plugins won't be part of the image, but will be persisted after the first (and only) manual installation.

One more observation for whoever stumbles here and plans to make a custom image like I did: it's probably a better idea to install python through pyenv instead of apt-get like I did, because the Debian base doesn't have the newer python versions in its repositories.

**Reply by msujew** [reply]
> So I suppose I could either mount the plugins folder on a persistent volume or bind-mount it on the host, then just proceed with the regular plugin installation through the UI, right? That way, the plugins won't be part of the image, but will be persisted after the first (and only) manual installation.

Yes, that's also completely feasible :)
